pkg/mcp/server/server.go

```diff
@@ -102,18 +115,37 @@ type connection struct {
 	watcher   Watcher
 }
 
-// New creates a new gRPC server that implements the Mesh Configuration Protocol (MCP).
-func New(watcher Watcher, supportedTypes []string) *Server {
+// New creates a new gRPC server that implements the Mesh Configuration Protocol (MCP). A nil authCheck
+// implies all incoming connections should be allowed.
+func New(watcher Watcher, supportedTypes []string, authChecker AuthChecker) *Server {
 	return &Server{
 		watcher:        watcher,
 		supportedTypes: supportedTypes,
+		authCheck:      authChecker,
 	}
 }
 
-func (s *Server) newConnection(stream mcp.AggregatedMeshConfigService_StreamAggregatedResourcesServer) *connection {
+func (s *Server) newConnection(stream mcp.AggregatedMeshConfigService_StreamAggregatedResourcesServer) (*connection, error) {
 	peerAddr := "0.0.0.0"
-	if peerInfo, ok := peer.FromContext(stream.Context()); ok {
+
+	peerInfo, ok := peer.FromContext(stream.Context())
+	if ok {
 		peerAddr = peerInfo.Addr.String()
+	} else {
+		log.Warnf("No peer info found on the incoming stream.")
+		peerInfo = nil
+	}
+
+	var authInfo credentials.AuthInfo
+	if peerInfo != nil {
+		authInfo = peerInfo.AuthInfo
+	}
+
+	if s.authCheck != nil {
```

Comment:
Should this fail if authCheck is required and peerInfo was missing? The only case where we should allow anonymous peers if if auth is disabled (testing only).

Reply:
I left it as the responsibility of the authCheck method. It may receive a "nil" AuthInfo, which it should handle appropriately.